File: Python/Algorithm/Is_Palindrome.py
# Complexity: O(len(s))

def pal(s):
    for i in range(len(s) // 2):
        if s[i] != s[-(i + 1)]: return False
    return True

# OR,

def pal(s):
    return all(s[i] == s[~i] for i in range(len(s)//2))

# Hashing Approach:

def pal(s, base = 29):
    sz = len(s); tmp = ord(s[0]) - ord('a') + 1
    temp = ord(s[sz - 1]) - ord('a') + 1; p = 1
    for i in range(1, sz):
        p = (p * base) % M
        tmp = (tmp + (ord(s[i]) - ord('a') + 1) * p) % M
    p = 1
    for i in range(sz - 2, -1, -1):
        p = (p * base) % M
        temp = (temp + (ord(s[i]) - ord('a') + 1) * p) % M
    return tmp == temp

File: CPP/Algorithm/Is_Palindrome.cpp
// Complexity: O(|s|)
// General Approach:

bool pal(auto &s) {
    for (int i = 0, j = s.size() - 1; i < j; ++i, --j) {
        if (s[i] != s[j]) return false;
    }
    return true;
}

// Hashing Approach:

bool pal(auto &s, int base = 29) {
    long long sz = s.size(), tmp = s[0] - 96, temp = s[sz - 1] - 96;
    for (int i = 1; i < sz; ++i) tmp = ((tmp * base) + (s[i] - 96)) % M;
    for (int i = sz - 2; i >= 0; --i) temp = ((temp * base) + (s[i] - 96)) % M;
    return (tmp == temp);
}